content/articles/How-to-start-up-with-Socket-Programming-with-Java/index.md

```diff
@@ -0,0 +1,315 @@
+- `public OutputStream getOutputStream()`- returns an OutputStream for sending data.
+- `public Socket accept()` - Waits for client connection (the program won't continue until the client is connected). When you connect, you'll get a socket object that you may use to communicate with the client. 
+- `void close()` - closes the server socket i.e. stops waiting for requests from clients.
+
+#### Full Server-Side Application
+```java
+package com. company; 
+ 
+import java.io.*;  
+import java.net.ServerSocket;  
+import java.net.Socket;  
+  
+public class Server {  
+  
+  public static void main(String[] args) throws IOException {   
+  Socket socket ;  
+  InputStreamReader inputStreamReader ;  
+  OutputStreamWriter outputStreamWriter ;  
+  BufferedReader bufferedReader ;  
+  BufferedWriter bufferedWriter ;  
+  ServerSocket serversocket ;  
+  
+serversocket = new ServerSocket(5000);  
+  
+ while (true) {  
+   try {  
+         
+         socket = serversocket.accept();  
+  
+inputStreamReader = new InputStreamReader(socket.getInputStream());  
+outputStreamWriter = new OutputStreamWriter(socket.getOutputStream());  
+bufferedReader = new BufferedReader(inputStreamReader);  
+bufferedWriter = new BufferedWriter(outputStreamWriter);  
+while (true)
+{  
+String msgFromClient = bufferedReader.readLine();  
+System.out.println("Client: " + msgFromClient);   
+bufferedWriter.write(" MSG Received"); 
+bufferedWriter.newLine();  
+bufferedWriter.flush();  
+if (msgFromClient.equalsIgnoreCase("BYE"))  
+  break;  
+  }  
+        socket.close();  
+        inputStreamReader.close();  
+        outputStreamWriter.close();  
+        bufferedReader.close();  
+        bufferedWriter.close();  
+  
+          } catch (IOException e) {  
+            e.printStackTrace();  
+          }  
+        }  
+    }  
+}
+```
+The output of this  program is:
+```
+Connected
+Client: Hey there 
+Client: I am doing this for the first time
+Client: BYE
+Closing connection
+```
+### Programming on the Client-side
+
+#### Establish a Socket Connection
+There are two ways to get a Socket instance:
+
+1. The server receives it as a return value of the accept method.
+2. You can use the following code to create a Socket:
+
+```java
+socket = new Socket("localhost", 5000);
+```
+Localhost is a domain name that redirects you to your computer, similar to google.com. It resolves to 127.0.0.1 as the IP address.
+A port number is in the second argument.
+
+We generate I/O streams using the socket object, just as we did before:
+
+```java
+inputStreamReader = new InputStreamReader(socket.getInputStream());  
+outputStreamWriter = new OutputStreamWriter(socket.getOutputStream());
+``` 
+Client-side, we need to use a Scanner object to get input from the user from the console/terminal. 
+
+```java
+Scanner scanner = new Scanner(System.in);
+```
+`System.in` specifies that we are getting the keyboard input.
+
+#### Methods used in Client Socket
```

Comment:
The methods you listed here are not all a part of the `Socket` class and this seems to suggest that. You also have listed some constructors here (i.e. `BufferReader()` and `BufferWriter()`). It doesn't make sense to list constructors as methods. Maybe change this header to something like:

> \#\#\#\# Methods and Objects used in the client-side program